✨ feat: 특정 알림 읽음 api 작성

yoo-sh-96 opened the pull request:
## ✅ PR 요약
- 관련 이슈 번호: #75
- 작업 요약: 특정 알림 읽음 api를 mock -> db기반으로 전환

## 📄 상세 내용
- [x] ORM 접근 managers로 전환
- [x] mock 기반에서 db기반으로 전환

## 📸 스크린샷 (선택)
<img width="1834" height="987" alt="스크린샷 2025-09-10 오전 11 46 22" src="https://github.com/user-attachments/assets/8d4a2970-44b7-4ecc-b2ed-e6df49e4b4bd" />

## 📝 기타 참고 사항
해당 알림만 읽음 처리 후 클릭하여 백링크로 이동에 대한 로직은 추후 서비스 레이어에서 django-signals와 selery task를 통해 구현 예정
## 🧪 PR Checklist
- [x] 커밋 메시지 컨벤션에 맞게 작성했습니다.
- [x] 변경 사항에 대한 테스트를 했습니다.(버그 수정/기능에 대한 테스트).

yoo-sh-96 committed:
✨ feat: 특정 알림 읽음 api 작성

mock -> db기반으로 전환
관련 이슈: #75

Meoyoug commented on apps/notifications/managers.py:
```diff
@@ -21,3 +21,8 @@ def list_queryset(self, user_id: int, status: str) -> QuerySet["Notification"]:
 
         return qs
+
+    def read_only_one_queryset(self, notification_id: int, user_id: int | None) -> int:
```
네이밍이 이상한것 같은데 하나의 쿼리셋을 읽음처리하는게 아니라 하나의 알림 객체를 읽음처리하는거니까 read라고만 지어도될것같아요
모든 알림 읽음처리는 read_all로하고

Meoyoug commented on apps/notifications/managers.py:
```diff
@@ -21,3 +21,8 @@ def list_queryset(self, user_id: int, status: str) -> QuerySet["Notification"]:
         return qs
+
+    def read_only_one_queryset(self, notification_id: int, user_id: int | None) -> int:
+        if user_id is None:
```
user_id가 None일 경우를 왜상정한걸까요..?

Meoyoug replied:
해당 알림을 수신한 유저가 읽음처리를 요청한 것인지 확인하는 건 비즈니스 로직에서 처리해야할 것 같아요!

yoo-sh-96 committed:
♻️ refactor: 비즈니스 로직 분리 및 테스트코드 추가

managers와 비즈니스 로직 분리 및 전체 읽음 테스트 코드 수정

관련 이슈: #75

yoo-sh-96 committed:
♻️ refactor: 타입 필터 추가 및 테스트 코드 추가

알림에 어떤 타입의 알림인지 표시해주는 태그가 있기에 타입 필터 추가

관련 이슈: #75

Meoyoug commented on apps/notifications/serializers.py:
```diff
@@ -39,3 +39,16 @@ class UnreadCountSerializer(serializers.Serializer[UnreadCountOut]):
+    is_read = serializers.BooleanField(required=False)
+    type = serializers.ChoiceField(choices=Notification.NotificationType.choices, required=False)
+
+    def validate(self, attrs: dict[str, Any]) -> dict[str, Any]:
```
required=False로 둔 필드들은 들어온 필드에 대해서만 검증을 할텐데 이 아래의 코드들은 유효하지 않을 것 같습니다

Meoyoug commented on apps/notifications/views.py:
```diff
@@ -39,12 +38,17 @@ class NotificationListView(ListAPIView[Notification]):
-
-        status_param = self.request.query_params.get("status", "all")
+        # 쿼리파라미터 검증
+        qp = NotificationListSerializer(data=self.request.query_params)
```
generic의 ListAPIView를 상속받았다면 쿼리파라미터검증은 get_queryset에서 하기보다 list메서드를 오버라이드하여 그 내부에서 하는것이 적합해보입니다

yoo-sh-96 committed:
🐛 fix: query serializer 수정 및 list 매서드 오버라이드

is_read를 CharField로 받은 뒤 validate_is_read()에서"true"/"false"만 허용 및 ListAPIView 사용으로 list매서드 오버라이드

관련 이슈: #75

yoo-sh-96 committed:
🎨 style: black 코드 포매터 실행

black 코드 포매터 실행

Meoyoug reviewed (approved):
고생하셨어요~
